In [171]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
# import chart_studio.plotly as py
import plotly.express as px
pd.set_option("display.max_columns", None)

### Read data
- Read the load data file
- Read the column decription file

In [316]:
data = pd.read_csv('../data/loan.csv', low_memory=False)
meta_data = pd.read_excel('../data/Data_Dictionary.xlsx')
meta_data.columns = ['column_name', 'description']

Getting the info of loan data using `data.info()` function

In [317]:
data.info(
    max_cols = 111,
    show_counts = True
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39717 entries, 0 to 39716
Data columns (total 111 columns):
 #    Column                          Non-Null Count  Dtype  
---   ------                          --------------  -----  
 0    id                              39717 non-null  int64  
 1    member_id                       39717 non-null  int64  
 2    loan_amnt                       39717 non-null  int64  
 3    funded_amnt                     39717 non-null  int64  
 4    funded_amnt_inv                 39717 non-null  float64
 5    term                            39717 non-null  object 
 6    int_rate                        39717 non-null  object 
 7    installment                     39717 non-null  float64
 8    grade                           39717 non-null  object 
 9    sub_grade                       39717 non-null  object 
 10   emp_title                       37258 non-null  object 
 11   emp_length                      38642 non-null  object 
 12   home_ownership  

### View of loan data
Top 5 rows of load data 

In [318]:
data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-11   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-11  Charged Off   
2           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65            0   
1                  bike    309xx         GA   1.00            0   
2  real estate business    606xx         IL   8.72            0   
3              personel    917xx         CA  20.00            0   
4              Personal    972xx         OR  17.94            0   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85               1                     NaN   
1           Apr-99               5                     NaN   
2           Nov-01               2                     NaN   
3           Feb-96               1                    35.0   
4           Jan-96               0                    38.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         3        0      13648     83.70%          9   
1                     NaN         3        0       1687      9.40%          4   
2                     NaN         2        0       2956     98.50%         10   
3                     NaN        10        0       5598        21%         37   
4                     NaN        15        0      27783     53.90%         38   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   f       0.00           0.00   5863.155187   
1                   f       0.00           0.00   1008.710000   
2                   f       0.00           0.00   3005.666844   
3                   f       0.00           0.00  12231.890000   
4                   f     524.06         524.06   3513.330000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16     

### Column analysis

Create new data frame for colum analysis where computed following data for each column
- Total number of missing values
- Total % of missing value
- Merged with description data
- Data type of column
- A shample column which store
    - If column is of int or float type then min and max value
    - If column is of object type the few shample values
- New categorical column for missing value categories depends on % of missing values
    - If 0 then VL (very low)
    - If b/w 1 - 10 then L(low)
    - If b/w 11 - 30 then M(medium)
    - If b/w 31 - 80 then H(high)
    - If b/w 81 - 100 then VH(very high)

In [319]:
column_analysis = pd.DataFrame(
    {'missing_value_percent':data.apply(lambda x:round((x.isna().sum()/x.shape[0])*100,2)),
     'total_of_nan':data.isna().sum(),
    'column_name':data.columns}
)


In [321]:
column_analysis.head()

missing_value_percent  total_of_nan      column_name
id                                 0.0             0               id
member_id                          0.0             0        member_id
loan_amnt                          0.0             0        loan_amnt
funded_amnt                        0.0             0      funded_amnt
funded_amnt_inv                    0.0             0  funded_amnt_inv

In [322]:
column_analysis.reset_index(inplace = True,drop = True)


In [323]:
column_analysis.head()

missing_value_percent  total_of_nan      column_name
0                    0.0             0               id
1                    0.0             0        member_id
2                    0.0             0        loan_amnt
3                    0.0             0      funded_amnt
4                    0.0             0  funded_amnt_inv

In [324]:
# Creating buckets according to percentage missing value
column_analysis['missing_category'] = pd.cut(x= column_analysis['missing_value_percent'], 
                                             bins= [-1.0, 0.0, 10.0, 30.0,80.0, 100.0],
                                            labels = ['Very Low (0% missing)','Low (1-10% missing)',
                                                  'Medium (11-30% missing)','High (31-80% missing)',
                                                  'Very High (80-100% missing)'])

In [325]:
def get_column_type(col_name):
    return data[col_name].dtype
column_analysis['dtype'] = column_analysis.column_name.apply(get_column_type)

In [326]:
def get_sample_value(col_name):
    col = column_analysis[column_analysis.column_name == col_name]
    dtype = col['dtype'].values[0]
    col_value = data[-data[col_name].isna()][col_name].values
    if (dtype == int or dtype == float) and col['missing_value_percent'].values[0] < 100:
        return F'{min(col_value)} | {max(col_value)}'
    elif dtype == object:
        _values = pd.Series(list(map(str,col_value))).unique()
        if(len(_values)/len(data) <= .15):
            return ' | '.join(_values)
        return ' | '.join(_values[:10])
    else:
        return None
column_analysis['sample_value'] = column_analysis.column_name.apply(get_sample_value)    

In [327]:
column_analysis.head()

missing_value_percent  total_of_nan      column_name  \
0                    0.0             0               id   
1                    0.0             0        member_id   
2                    0.0             0        loan_amnt   
3                    0.0             0      funded_amnt   
4                    0.0             0  funded_amnt_inv   

        missing_category    dtype   sample_value  
0  Very Low (0% missing)    int64           None  
1  Very Low (0% missing)    int64           None  
2  Very Low (0% missing)    int64           None  
3  Very Low (0% missing)    int64           None  
4  Very Low (0% missing)  float64  0.0 | 35000.0

### Calculated % of column under each category

Grouping rows according to missing_category column using pivot table
- create column from pivot table index 
- drop index
- plot pie chart

In [328]:
missing_category_percentage = pd.pivot_table(column_analysis,index = 'missing_category',values = 'missing_value_percent',aggfunc = lambda x:len(x)/len(column_analysis)*100)

In [329]:
missing_category_percentage['missing_category'] = missing_category_percentage.index


In [330]:
missing_category_percentage.reset_index(inplace = True,drop = True)

In [331]:
missing_category_percentage

missing_value_percent             missing_category
0              38.738739        Very Low (0% missing)
1               9.009009          Low (1-10% missing)
2               1.801802        High (31-80% missing)
3              50.450450  Very High (80-100% missing)

### Ploting missing % of columns values category wise

Some intresting insite found
- 50% of columns have 100% missing value
- 1.8% of columns have 31 - 80% missing values
- Only 38.7% of columns have 0% missing values

So decided to go with only those rows which have less then 10% missing value

![Visualization](./newplot.png)

In [332]:
fig = px.pie(
    missing_category_percentage, 
    values='missing_value_percent', 
    names='missing_category',
    title='Percentage of column under mssing categories(depends on % on value missing)',
    hole=.5,
    color_discrete_sequence=px.colors.sequential.Agsunset_r,
    labels={'missing_category':'Category', 'column_percent': 'Column Percentage'},
)
fig.show()

In [ ]:
# column_analysis.set_index('column_name', inplace=True)

### Saving the column analysis data for future review

In [334]:
# column_analysis.to_excel('../data/explore_data.xlsx')
column_analysis.to_csv('../data/explore_data.csv')

In [335]:
missing_category_percentage

missing_value_percent             missing_category
0              38.738739        Very Low (0% missing)
1               9.009009          Low (1-10% missing)
2               1.801802        High (31-80% missing)
3              50.450450  Very High (80-100% missing)

### dropping columns with more than 30 percent null values

In [337]:
column_values_with_more_than_30_percent_missing = list(column_analysis[column_analysis['missing_category'].isin(['High (31-80% missing)','Very High (80-100% missing)'])]['column_name'].values)

In [338]:
column_values_with_more_than_30_percent_missing

['desc',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'next_pymnt_d',
 'mths_since_last_major_derog',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'tot_coll_amt',
 'tot_cur_bal',
 'open_acc_6m',
 'open_il_6m',
 'open_il_12m',
 'open_il_24m',
 'mths_since_rcnt_il',
 'total_bal_il',
 'il_util',
 'open_rv_12m',
 'open_rv_24m',
 'max_bal_bc',
 'all_util',
 'total_rev_hi_lim',
 'inq_fi',
 'total_cu_tl',
 'inq_last_12m',
 'acc_open_past_24mths',
 'avg_cur_bal',
 'bc_open_to_buy',
 'bc_util',
 'mo_sin_old_il_acct',
 'mo_sin_old_rev_tl_op',
 'mo_sin_rcnt_rev_tl_op',
 'mo_sin_rcnt_tl',
 'mort_acc',
 'mths_since_recent_bc',
 'mths_since_recent_bc_dlq',
 'mths_since_recent_inq',
 'mths_since_recent_revol_delinq',
 'num_accts_ever_120_pd',
 'num_actv_bc_tl',
 'num_actv_rev_tl',
 'num_bc_sats',
 'num_bc_tl',
 'num_il_tl',
 'num_op_rev_tl',
 'num_rev_accts',
 'num_rev_tl_bal_gt_0',
 'num_sats',
 'num_tl_120dpd_2m',
 'num_tl_30dpd',
 'num_tl_90g_dpd_24m',
 'num_tl_op_pa

In [305]:
data_clean = data.drop(column_values_with_more_than_30_percent_missing,axis = 1)

In [315]:
data_clean.isna().sum()

id                               0
member_id                        0
loan_amnt                        0
funded_amnt                      0
funded_amnt_inv                  0
term                             0
int_rate                         0
installment                      0
grade                            0
sub_grade                        0
emp_title                     2459
emp_length                    1075
home_ownership                   0
annual_inc                       0
verification_status              0
issue_d                          0
loan_status                      0
pymnt_plan                       0
url                              0
purpose                          0
title                           11
zip_code                         0
addr_state                       0
dti                              0
delinq_2yrs                      0
earliest_cr_line                 0
inq_last_6mths                   0
open_acc                         0
pub_rec             

In [339]:
data_clean.to_csv('../data/nan_drop_data.csv')

In [340]:
data_clean.shape

(39717, 53)

### Analysis 14-  24 column


In [426]:
def column_meaning(column_name):
    '''Returns meainng of the column from data deoctionary provided
    '''
    meta_data = pd.read_excel('../data/Data_Dictionary.xlsx')
    meta_data.columns = ['column_name', 'description']
    return meta_data[meta_data['column_name']==column_name]['description'].values
    

In [341]:
df = data_clean.iloc[:,14:25]

In [415]:
df.isna().sum()

verification_status     0
issue_d                 0
loan_status             0
pymnt_plan              0
url                     0
purpose                 0
title                  11
zip_code                0
addr_state              0
dti                     0
delinq_2yrs             0
month                   0
year                    0
dtype: int64

In [342]:
df.head()

verification_status issue_d  loan_status pymnt_plan  \
0            Verified  Dec-11   Fully Paid          n   
1     Source Verified  Dec-11  Charged Off          n   
2        Not Verified  Dec-11   Fully Paid          n   
3     Source Verified  Dec-11   Fully Paid          n   
4     Source Verified  Dec-11      Current          n   

                                                 url         purpose  \
0  https://lendingclub.com/browse/loanDetail.acti...     credit_card   
1  https://lendingclub.com/browse/loanDetail.acti...             car   
2  https://lendingclub.com/browse/loanDetail.acti...  small_business   
3  https://lendingclub.com/browse/loanDetail.acti...           other   
4  https://lendingclub.com/browse/loanDetail.acti...           other   

                  title zip_code addr_state    dti  delinq_2yrs  
0              Computer    860xx         AZ  27.65            0  
1                  bike    309xx         GA   1.00            0  
2  real estate business    606xx         IL   8.72            0  
3              personel    917xx         CA  20.00            0  
4              Personal    972xx         OR  17.94            0

## 1. issue_d
- Creating 2 new columns for month and year
- frequency curve for month across different years

In [344]:
df.issue_d.unique()

array(['Dec-11', 'Nov-11', 'Oct-11', 'Sep-11', 'Aug-11', 'Jul-11',
       'Jun-11', 'May-11', 'Apr-11', 'Mar-11', 'Feb-11', 'Jan-11',
       'Dec-10', 'Nov-10', 'Oct-10', 'Sep-10', 'Aug-10', 'Jul-10',
       'Jun-10', 'May-10', 'Apr-10', 'Mar-10', 'Feb-10', 'Jan-10',
       'Dec-09', 'Nov-09', 'Oct-09', 'Sep-09', 'Aug-09', 'Jul-09',
       'Jun-09', 'May-09', 'Apr-09', 'Mar-09', 'Feb-09', 'Jan-09',
       'Dec-08', 'Nov-08', 'Oct-08', 'Sep-08', 'Aug-08', 'Jul-08',
       'Jun-08', 'May-08', 'Apr-08', 'Mar-08', 'Feb-08', 'Jan-08',
       'Dec-07', 'Nov-07', 'Oct-07', 'Sep-07', 'Aug-07', 'Jul-07',
       'Jun-07'], dtype=object)

In [375]:
df['month']= pd.DatetimeIndex(pd.to_datetime(df['issue_d'],format = '%b-%y')).month_name()
df['year']= pd.DatetimeIndex(pd.to_datetime(df['issue_d'],format = '%b-%y')).year

In [376]:
df.head()

verification_status issue_d  loan_status pymnt_plan  \
0            Verified  Dec-11   Fully Paid          n   
1     Source Verified  Dec-11  Charged Off          n   
2        Not Verified  Dec-11   Fully Paid          n   
3     Source Verified  Dec-11   Fully Paid          n   
4     Source Verified  Dec-11      Current          n   

                                                 url         purpose  \
0  https://lendingclub.com/browse/loanDetail.acti...     credit_card   
1  https://lendingclub.com/browse/loanDetail.acti...             car   
2  https://lendingclub.com/browse/loanDetail.acti...  small_business   
3  https://lendingclub.com/browse/loanDetail.acti...           other   
4  https://lendingclub.com/browse/loanDetail.acti...           other   

                  title zip_code addr_state    dti  delinq_2yrs     month  \
0              Computer    860xx         AZ  27.65            0  December   
1                  bike    309xx         GA   1.00            0  December   
2  real estate business    606xx         IL   8.72            0  December   
3              personel    917xx         CA  20.00            0  December   
4              Personal    972xx         OR  17.94            0  December   

   year  
0  2011  
1  2011  
2  2011  
3  2011  
4  2011

In [406]:
# plt.figure(figsize = (20,10))
# # plt.title("Trend of data")

# sns.barplot(x =df.month.value_counts(),y = df.month.value_counts().index).set(title='Points by Team')

In [442]:

# tempdf = pd.DataFrame({'month':df.month.value_counts().index,'count':df.month.value_counts()})
# fig = px.bar(tempdf,x ='month',y = 'count',color='month',title ="Monthly loan funded accross years")

# fig.show()

In [440]:
def univariate(df,column_name,title):
    tempdf = pd.DataFrame({'category':df[column_name].value_counts().index,'count':df[column_name].value_counts()})
    fig = px.bar(tempdf,x ='category',y = 'count',color='category',title =title)

    fig.show()
    

In [441]:
univariate(df,'month',"Monthly loan funded accross years")

In [443]:

# tempdf = pd.DataFrame({'year':df.year.value_counts().index,'count':df.year.value_counts()})
# fig = px.bar(tempdf,x ='year',y = 'count',color='year',title ="Yearly loan funded accross months")

# fig.show()
univariate(df,'year',"Yearly loan funded accross months")

In [417]:
df.year.value_counts()

2011    21656
2010    11532
2009     4716
2008     1562
2007      251
Name: year, dtype: int64

In [427]:
column_meaning(column_name = 'issue_d')

array(['The month which the loan was funded'], dtype=object)

## issue_d: Findings
- the loans funded across differnt months indicate a linear trend as the year progress so are the loans funded
- most loans were funded in year 2011
- as this a column filled after loan is approved it should be dropped

## 2. verification_status

In [418]:
df.columns

Index(['verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url',
       'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs',
       'month', 'year'],
      dtype='object')

In [428]:
column_meaning("verification_status")

array(['Indicates if income was verified by LC, not verified, or if the income source was verified'],
      dtype=object)

In [432]:
col = 'verification_status'
df[col].unique(),df[col].isna().sum()


(array(['Verified', 'Source Verified', 'Not Verified'], dtype=object), 0)

In [444]:
univariate(df,col,"verification_status frequency")

## verification_status: Findings:
- bivariate analysis should be done to see how this categry is related to loan defaulter
- no nan values

## 3. loan_status
- Target column
    - Fully paid: Applicant has fully paid the loan (the principal and the interest rate)

    - Current: Applicant is in the process of paying the instalments, i.e. the tenure of the loan is not yet completed. These candidates are not labelled as 'defaulted'.

    - Charged-off: Applicant has not paid the instalments in due time for a long period of time, i.e. he/she has defaulted on the loan 

col = 'loan_status'

In [447]:
df[col].isna().sum()

0

In [450]:
df[col].unique()

array(['Fully Paid', 'Charged Off', 'Current'], dtype=object)

In [451]:
column_meaning(col)

array(['Current status of the loan'], dtype=object)

In [456]:
df[col].value_counts()/df[col].shape[0]

Fully Paid     0.829620
Charged Off    0.141677
Current        0.028703
Name: loan_status, dtype: float64

In [465]:
# fig = px.pie(df, values='tip', names='day')
fig = px.pie(df, names=col, title='Distribution of loan status in dataset')
fig.show()

## loan_status: findings
- target column
- no missing value
- data is skewed ie around 83 percent of the data is for people who repayed loan so the Defaulter class which is the main target of this for the company is in minority

In [467]:
df.columns

Index(['verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url',
       'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs',
       'month', 'year'],
      dtype='object')

## 4. pymnt_plan

In [468]:
col = 'pymnt_plan'

In [469]:
df[col].isna().sum()

0

In [470]:
df[col].dtypes

dtype('O')

In [472]:
df[col].unique()

array(['n'], dtype=object)

In [473]:
column_meaning(col)

array(['Indicates if a payment plan has been put in place for the loan'],
      dtype=object)

## pymnt_plan: findings
- single value column, contains only 1 value - 'n'
- wont be helpful in analysis hence should be dropped

## 5. url

In [474]:
col = 'url'
df[col].dtype

dtype('O')

In [475]:
df[col].isna().sum()

0

In [482]:
df[col].iloc[0:5].values

array(['https://lendingclub.com/browse/loanDetail.action?loan_id=1077501',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=1077430',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=1077175',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=1076863',
       'https://lendingclub.com/browse/loanDetail.action?loan_id=1075358'],
      dtype=object)

In [478]:
column_meaning(col)

array(['URL for the LC page with listing data.'], dtype=object)

In [480]:
data_clean.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-11   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-11  Charged Off   
2           RENT     12252.0        Not Verified  Dec-11   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-11   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-11      Current   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

          purpose                 title zip_code addr_state    dti  \
0     credit_card              Computer    860xx         AZ  27.65   
1             car                  bike    309xx         GA   1.00   
2  small_business  real estate business    606xx         IL   8.72   
3           other              personel    917xx         CA  20.00   
4           other              Personal    972xx         OR  17.94   

   delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0            0           Jan-85               1         3        0      13648   
1            0           Apr-99               5         3        0       1687   
2            0           Nov-01               2         2        0       2956   
3            0           Feb-96               1        10        0       5598   
4            0           Jan-96               0        15        0      27783   

  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0     83.70%          9                   f       0.00           0.00   
1      9.40%          4                   f       0.00           0.00   
2     98.50%         10                   f       0.00           0.00   
3        21%         37                   f       0.00           0.00   
4     53.90%         38                   f     524.06         524.06   

    total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0   5863.155187          5833.84          5000.00         863.16   
1   1008.710000          1008.71           456.46         435.17   
2   3005.666844          3005.67          2400.00         605.67   
3  12231.890000         12231.89         10000.00        2214.92   
4   3513.330000          3513.33          2475.94        1037.39   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00       Jan-15   
1                0.00      117.08                     1.11       Apr-13   
2                0.00        0.00                     0.00       Jun-14   
3               16.97        0.00                     0.00       Jan-15   
4                0.00        0.00                     0.00       May-16   

   last_p

In [485]:
data_clean['id'].dtype

dtype('int64')

In [489]:
temp = df[col].apply(lambda x:int(x.split('=')[-1])) == data_clean['id']

In [492]:
temp.sum() == data_clean.shape[0]

True

## url: findings
- conatins redundant information to loan_id
- should be dropped

In [502]:
def column_meaning(column_name):
    '''Returns meainng of the column from data deoctionary provided
    '''
    meta_data = pd.read_excel('../data/Data_Dictionary.xlsx')
    meta_data.columns = ['column_name', 'description']
    return meta_data[meta_data['column_name']==column_name]['description'].values
def insights(column_name,df):
#     ta_data = pd.read_excel('../data/Data_Dictionary.xlsx')
#     meta_data.columns = ['column_name', 'description']
#     des = meta_data[meta_data['column_name']==column_name]['description'].values
    typ = df[column_name].dtype
    null_val = df[column_name].value_counts
#     print("description: ",des)
    print("type: ",typ)
    print("null value count: ",null_val)
    

## 6. purpose

In [493]:
df.columns

Index(['verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url',
       'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs',
       'month', 'year'],
      dtype='object')

In [504]:
col = 'purpose'
column_meaning(col)

array(['A category provided by the borrower for the loan request. '],
      dtype=object)

In [506]:
df[col].dtype

dtype('O')

In [507]:
df[col].isna().sum()

0

In [508]:
df[col].unique()

array(['credit_card', 'car', 'small_business', 'other', 'wedding',
       'debt_consolidation', 'home_improvement', 'major_purchase',
       'medical', 'moving', 'vacation', 'house', 'renewable_energy',
       'educational'], dtype=object)

In [509]:
fig = px.pie(df, names=col, title='Distribution of loan status in dataset')
fig.show()

## purpose: findings
- important column
- use for segmented univariate

## 7. title

In [510]:
df.columns

Index(['verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url',
       'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs',
       'month', 'year'],
      dtype='object')

In [512]:
col = 'title'
df[col].isna().sum()

11

In [521]:
df[col].unique()

array(['Computer', 'bike', 'real estate business', ...,
       'Retiring credit card debt', 'MBA Loan Consolidation', 'JAL Loan'],
      dtype=object)

## title: findings
- too many unique string values hence dont add meaning
- should be dropped

In [522]:
col = 'zip_code'

In [523]:
df[col].unique()

array(['860xx', '309xx', '606xx', '917xx', '972xx', '852xx', '280xx',
       '900xx', '958xx', '774xx', '853xx', '913xx', '245xx', '951xx',
       '641xx', '921xx', '067xx', '890xx', '770xx', '335xx', '799xx',
       '605xx', '103xx', '150xx', '326xx', '564xx', '141xx', '080xx',
       '330xx', '974xx', '934xx', '405xx', '946xx', '445xx', '850xx',
       '604xx', '292xx', '088xx', '180xx', '029xx', '700xx', '010xx',
       '441xx', '104xx', '061xx', '616xx', '947xx', '914xx', '765xx',
       '980xx', '017xx', '752xx', '787xx', '077xx', '540xx', '225xx',
       '440xx', '437xx', '559xx', '912xx', '325xx', '300xx', '923xx',
       '352xx', '013xx', '146xx', '074xx', '786xx', '937xx', '331xx',
       '115xx', '191xx', '114xx', '908xx', '902xx', '992xx', '750xx',
       '950xx', '329xx', '226xx', '614xx', '802xx', '672xx', '083xx',
       '100xx', '926xx', '931xx', '712xx', '060xx', '707xx', '342xx',
       '895xx', '430xx', '919xx', '996xx', '891xx', '935xx', '801xx',
       '928xx', '233